File: json-schema/research/format_materialize_clock/runner.rb
# Probe: MATERIALIZE model (B) in Ruby via stdlib DateTime/Date/Time.
# PROSPECTIVE target. Parse each validated wire string, re-serialize to the
# CANONICAL form, emit the bytes. ruby runner.rb corpus.json
require "json"
require "date"
require "time"

ENGINE = "ruby"

def emit(o)
  puts JSON.generate(o)
end

# date-time -> UTC, ms, "YYYY-MM-DDTHH:MM:SS.mmmZ".
# NOTE: DateTime.rfc3339 CLAMPS leap :60 -> :59 silently (prior finding).
def canon_datetime(wire)
  dt = DateTime.rfc3339(wire.upcase) # raises on missing offset; CLAMPS :60->:59
  u = dt.new_offset(0)
  # truncate to ms
  ms = (u.sec_fraction * 1000).to_i
  format("%04d-%02d-%02dT%02d:%02d:%02d.%03dZ",
         u.year, u.month, u.day, u.hour, u.min, u.sec, ms)
end

def canon_date(wire)
  d = Date.iso8601(wire)
  format("%04d-%02d-%02d", d.year, d.month, d.day)
end

# time -> Ruby has NO time-of-day-only type. Time.parse fabricates today's date.
def canon_time(_wire)
  raise "UNSUPPORTED: Ruby has no time-of-day-only type"
end

def run(rows, fmt, fn)
  rows.each do |r|
    begin
      emit({ id: r["id"], engine: ENGINE, format: fmt, canonical: fn.call(r["wire"]), err: "" })
    rescue => e
      emit({ id: r["id"], engine: ENGINE, format: fmt, canonical: "", err: "#{e.class}: #{e.message}" })
    end
  end
end

c = JSON.parse(File.read(ARGV[0]))
run(c["date-time"], "date-time", method(:canon_datetime))
run(c["date"], "date", method(:canon_date))
run(c["time"], "time", method(:canon_time))

File: json-schema/research/format_materialize_clock/runner.mjs
// Probe: MATERIALIZE model (B) in JS/TS. The only stdlib temporal type is
// `Date` (a UTC instant, millisecond floor, offset LOST). date-only and
// time-only have NO stdlib type. Parse each validated wire string, re-serialize
// to the CANONICAL form, emit the bytes.  node runner.mjs corpus.json
import { readFileSync } from "node:fs";

const ENGINE = "js";
const emit = (o) => console.log(JSON.stringify(o));

// date-time via Date. Date.parse is LAX (accepts calendar-invalid, missing
// offset) but here we only feed strings already validated by the pinned check,
// so the ONLY question is round-trip bytes. Date is UTC + ms floor.
function canonDateTime(wire) {
  const ms = Date.parse(wire);
  if (Number.isNaN(ms)) throw new Error("Date.parse -> NaN");
  const d = new Date(ms);
  // toISOString() is always "YYYY-MM-DDTHH:mm:ss.sssZ" (UTC, exactly 3 frac, Z)
  return d.toISOString();
}

// date-only: JS has NO date type. new Date("2021-06-15") is a UTC instant at
// midnight. We CAN recover Y-M-D from the UTC fields since the input midnight
// is interpreted as UTC. But this is materializing as a full instant, not a
// date. Emit Y-M-D from UTC parts to test byte-equality; flag it lossy.
function canonDate(wire) {
  const ms = Date.parse(wire); // "2021-06-15" -> UTC midnight
  if (Number.isNaN(ms)) throw new Error("Date.parse -> NaN");
  const d = new Date(ms);
  const y = String(d.getUTCFullYear()).padStart(4, "0");
  const mo = String(d.getUTCMonth() + 1).padStart(2, "0");
  const da = String(d.getUTCDate()).padStart(2, "0");
  return `${y}-${mo}-${da}`;
}

// time-only: JS has NO time type. new Date("12:30:45") -> Invalid Date.
// Cannot materialize as a language construct at all. Report unsupported.
function canonTime(_wire) {
  throw new Error("UNSUPPORTED: JS has no time-of-day type");
}

function run(rows, fmt, fn) {
  for (const r of rows) {
    try {
      emit({ id: r.id, engine: ENGINE, format: fmt, canonical: fn(r.wire), err: "" });
    } catch (e) {
      emit({ id: r.id, engine: ENGINE, format: fmt, canonical: "", err: String(e.message || e) });
    }
  }
}

const c = JSON.parse(readFileSync(process.argv[2], "utf8"));
run(c["date-time"], "date-time", canonDateTime);
run(c["date"], "date", canonDate);
run(c["time"], "time", canonTime);
